Issue all non-upload storage-quota-update events from LiteElement

tw4l opened the pull request:
Follow up to #1060

tw4l committed:
Send storage-quota-update event from LiteElement on 403 storage_quota_reached

tw4l committed:
Parse JSON earlier

tw4l committed:
Try to get detail from body earlier

SuaYoo committed:
remove try-catch

SuaYoo committed:
stop propagation

SuaYoo committed:
adjust api handler

SuaYoo committed:
move message to org component

SuaYoo committed:
update error messages

SuaYoo committed:
move alert

SuaYoo committed:
update upload message

SuaYoo committed:
remove unused import

SuaYoo committed:
remove first update test

SuaYoo commented:
@tw4l I added the following:

1. More specific toast notification error messages to the action being attempted

2. Single dismissable global banner shown when org storage is reached:

<img width="1439" alt="Screenshot 2023-09-07 at 5 42 48 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/4672952/240dfe93-8ebd-41f9-aeee-a494a187787c">


3. Removed check for storage quota reached in `runNow`, since buttons are disabled in UI, and errors handled if request fails.

4. Allow creating new workflow

Tested only with frontend mocks, not real backend values. What's the best way to test storage quotas locally?

SuaYoo commented on frontend/src/components/file-uploader.ts:
```diff
@@ -459,13 +459,13 @@ export class FileUploader extends LiteElement {
     } catch (err: any) {
       if (err === ABORT_REASON_USER_CANCEL) {
-        console.debug("Fetch crawls aborted to user cancel");
+        console.debug("Upload aborted to user cancel");
```
unrelated fix!

SuaYoo committed:
update message

SuaYoo committed:
show quota on load

SuaYoo committed:
allow creating new workflow

SuaYoo committed:
handle missing body json

ikreymer commented:
Deployed to dev server for easier testing.

SuaYoo committed:
disable other run crawls

tw4l commented:
> @tw4l I added the following:
> 
>     1. More specific toast notification error messages to the action being attempted
> 
>     2. Single dismissable global banner shown when org storage is reached:
> 
>     3. Removed check for storage quota reached in `runNow`, since buttons are disabled in UI, and errors handled if request fails.
> 
>     4. Allow creating new workflow
> 
> 
> Tested only with frontend mocks, not real backend values. What's the best way to test storage quotas locally?

Nice!

Can test locally same way locally or on dev - from the orgs list, click on the gear icon next to an org and set its storage quota low, like 1 or 2 GB, upload some big WACZs that get it close, and then run a crawl or upload that'll set it over the limit. I have some large WACZ files I've been using to test, lmk on Discord if you'd like me to share.

tw4l commented:
Testing on dev (in "tessa" org), I see that the toast notifications still appear to be in the lower-right corner, rather than at the top. Additionally, everything works as expected when an upload is what puts it over the storage quota but when it's a crawl that does it, the update doesn't seem to happen in the frontend until you attempt to run another crawl or refresh the browser. To replicate, go ahead and run the one workflow in that org - there are about 960 MB of uploads in that org and a 1 GB storage quota set, so the crawl will put it over.

SuaYoo commented:
> Testing on dev (in "tessa" org), I see that the toast notifications still appear to be in the lower-right corner, rather than at the top.

I left the toast notification for the specific failed action in its usual place, so that the user can directly correlate an action with the error. The banner up top is an additional alert that persists until user action if storage is full, for instance on page load.

> when it's a crawl that does it, the update doesn't seem to happen in the frontend until you attempt to run another crawl

It appears to be a backend issue, the `/run` endpoint is returning `200` with body `started: crawl_id` instead of an error code or `storageQuotaReached`.

tw4l commented:
> It appears to be a backend issue, the `/run` endpoint is returning `200` with body `started: crawl_id` instead of an error code or `storageQuotaReached`.

Ah right, that makes sense because the `/run` endpoint does the quota check and returns when the crawl starts, and at that point the storage quota hasn't yet been exceeded. Maybe we should add `storageQuotaReached` in the `replay.json` endpoint so that we catch it after the crawl completes?

tw4l committed:
Set storageQuotaReached on crawl finish

- Add storageQuotaReached to archived item replay.json
- Modify LiteElement to check for storageQuotaReached on GET requests

tw4l commented:
@SuaYoo I've made some changes in the latest commit that makes the frontend responsive to the storage quota on crawl finish. Let me know what you think.

SuaYoo commented:
@tw4l nice, lgtm!

@ikreymer this should be ready for re-review, with both Tessa and my changes in.

ikreymer reviewed (approved):
Nice! Only question is if we replay.json is correct place to include storage quota, but that seems to be a commonly refreshed endpoint, perhaps revisit later if we have WS-based updates.

Tested with running a crawl that exceeds quota and saw the UI update w/o refresh
